src/autodiff_model.jl

```diff
@@ -39,31 +51,14 @@ and ``c_L`` and ``c_U`` should be passed, otherwise the problem is ill-formed.
 For equality constraints, the corresponding index of ``c_L`` and ``c_U`` should be the
 same.
 """
-mutable struct ADNLPModel <: AbstractNLPModel
-  meta :: NLPModelMeta
-
-  counters :: Counters
-
-  # Functions
-  f
-  c
-end
+function ADNLPModel(f, x0::AbstractVector{T}; nvar :: Integer = length(x0), ncon :: Integer = 0,
+                    lvar::AbstractVector = fill(T(-Inf), nvar), uvar::AbstractVector = fill(T(Inf), nvar),
```

Comment:
Don't we want `lvar = fill!(similar(x0), T(-Inf))`, etc.? Otherwise, `fill()` creates standard vectors.

Reply:
I'm not sure how to handle lcon and ucon though. We could make it mandatory to input initial multipliers, so we could use the same syntax?!

Reply:
:+1: on the first comment

On the second, I rather not having mandatory multipliers. Also, having mandatory multipliers would still break the current syntax.

To break some of syntax we could have
- `lcon = T[]`
- `ucon = T[]`
- `ncon = max(length(lcon), length(ucon))`
- `y0 = zeros(T, ncon)`
in that order in the keywords. Which is similar to what we already had. So `lcon` and `ucon` need to be defined for constraints, but `y0` has default.

I think I forgot why we need `ncon`.

Reply:
I'm not sure I understand. `lcon = T[]` and `y0 = zeros(T, ncon)` both create standard vectors. If `length(lcon)` is not the same as `length(ucon)`, it should be an error.

I'm trying to find a way for `lcon` and `ucon` to be of the same type as `c(x)`, which should be the same type as `y0`. I suggested `y0` because we could then use `lcon = fill!(similar(y0), typeof(y0)(-Inf))`, etc.

Reply:
In the unconstrained case we'd have `c(x) = T[]`, `lcon = T[]` and `ucon = T[]` (and therefore `y0 = zeros(T, 0) = T[]`). If the user defines constraints, such as `c(x) = CuArray[...]` (or however is done), then he should set `lcon` and `ucon` correctly as well -- inside the code we'll still `@lencheck ncon lcon ucon y0` to make sure that `lcon` and `ucon` are  the same size.
We can use `y0 = fill!(similar(lcon), 0)`, then.

This means that `ncon` is kind of unused, though.

Reply:
Ok. I have the feeling that we should make those arguments positional instead of kwargs then:
* `ADNLPModel(f, x0)`: unconstrained;
* `ADNLPModel(f, x0, lvar, uvar)`: bound-constrained (here, `lvar` and `uvar` can be positional and have default values;
* `ADNLPModel(f, x0, c, lcon, ucon)`: general constraints, no bounds; this could have a kwarg for `y0` with default value `fill!(similar(lcon), 0)`, but I don't see how to give `lcon` and `ucon` default values other than by making `y0` mandatory;
* `ADNLPModel(f, x0, lvar, uvar, c, lcon, ucon)`: general.

I hate to say this but... perhaps we should do the same across the board?!

Reply:
Ok, these are the ones implemented, but should we remove the option for creating with keywords altogether?

I don't understand what do you mean with across the board.

Reply:
> I don't understand what do you mean with across the board.

For all NLPModels (not in this PR!).

Reply:
I guess that makes sense to QuadraticModels, but I don't see it in other models.

Reply:
In CUTEst and AmplNLReader, it doesn't make sense, but it does in the more general AD model that I'm writing. And we should keep the interface consistent.

Reply:
Sure, that makes sense. I think that your suggestion for positional arguments is great, but some clarifications:
- should we have a (non-documented/internal?) model with keywords arguments?  For instance so the other methods call this one.
- related, if I try `ADNLPModel(f, x0, lvar=lvar, uvar=uvar)`, should it error?
- same for other variants.

Reply:
We need to make an executive decision because from what I see, keyword args won't let us properly define lcon/ucon/y0 so they have the right AbstractVector type. So maybe we shouldn't allow keyword args at all in the constructor. Is there any disadvantage to that?

Reply:
I've made some updates and from the looks of it, I think disallowing keyword arguments should be the way to deal with it. I haven't found any disadvantages.